New CAEP event - Risk level change event (#205)
openid/sharedsignals: openid-caep-1_0.md

~~~

## Risk Level Change {#risk-level-change}
Event Type URI:

`https://schemas.openid.net/secevent/caep/event-type/risk-level-change`

A vendor may deploy mechanisms to gather and analyze various signals associated with subjects such as users, devices, etc. These signals, which can originate from diverse channels and methods beyond the scope of this event description, are processed to derive an abstracted risk level representing the subject's current threat status.

The Risk Level Change event is employed by the Transmitter to communicate any modifications in a subject's assessed risk level at the time indicated by the `event_timestamp` field in the Risk Level Change event. The Transmitter may generate this event to indicate:

* User's risk has changed due to potential suspecious access from unknown destination, password compromise, addition of strong authenticator or other reasons.
* Device's risk has changed due to installation of unapproved software, connection to insecure pheripheral device, encryption of data or other reasons.
* Any other subject's risk changes due to variety of reasons.

### Event Specific Claims {#risk-level-change-event-specific-claims}

risk_reason

> : REQUIRED, JSON string indicates the reason that contributed to the risk level changes by the Transmitter.

Reviewer: ```suggestion
: REQUIRED, JSON object: a localizable message indicating the reason that contributed to the risk level changes by the Transmitter. The object MUST contain one or more key/value pairs, with a BCP47 {{RFC5646}} language tag as the key and the locale-specific administrative message as the value.
```

Reviewer: Why is this REQUIRED? It seems like it ought to be an optional field

Author: `risk_reason` is marked required since, without this, there is no benefit for the receiver to process such an event. Transmitter updating the risk will be a black box process and hinder the visibility.

Also, it may not make sense to localize the message as this is intended for the receiver to make sense why Tx decided to update the risk. This could become ENUM between Tx-Rx. (Happy to update the description if that clarifies)

User-visible messages would be part of `reason_admin` and `reason_user` which could be localized

current_level 

> : REQUIRED, JSON string indicates the current level of the risk for the subject. Value MUST be one of LOW, MEDIUM, HIGH

Reviewer: ```suggestion
: REQUIRED, string: indicates the current level of the risk for the subject. Value MUST be one of LOW, MEDIUM, HIGH
```

Author: Following the same convention of the other CAEP events

previous_level 

> : OPTIONAL, JSON string indicates the previously known level of the risk for the subject. Value MUST be one of LOW, MEDIUM, HIGH. If the Transmitter omits this value, the Receiver MUST assume that the previous assurance level is unknown to the Transmitter.

Reviewer: ```suggestion
: OPTIONAL, string: indicates the previously known level of the risk for the subject. Value MUST be one of LOW, MEDIUM, HIGH. If the Transmitter omits this value, the Receiver MUST assume that the previous assurance level is unknown to the Transmitter.
```

Reviewer: I think the second statement ought to be a SHOULD:

> If the Transmitter omits this value, the Receiver SHOULD assume that the previous assurance level is unknown to the Transmitter.

Author: Following the same convention of the other CAEP events

    }}
~~~

## Risk Level Change {#risk-level-change}
Event Type URI:

`https://schemas.openid.net/secevent/caep/event-type/risk-level-change`

A vendor may deploy mechanisms to gather and analyze various signals associated with subjects such as users, devices, etc. These signals, which can originate from diverse channels and methods beyond the scope of this event description, are processed to derive an abstracted risk level representing the subject's current threat status.

> The Risk Level Change event is employed  by the Trasnmitter to communicate any modifications in a subject's assessed risk level at the time indicated by the `event_timestamp` field in the Risk Level Change event. The Transmitter may generate this event to indicate:

Reviewer: ```suggestion
The Risk Level Change event is employed by the Transmitter to communicate any modifications in a subject's assessed risk level at the time indicated by the `event_timestamp` field in the Risk Level Change event. The Transmitter may generate this event to indicate:
```

* User's risk has changed due to potential suspecious access from unknown destination, password compromise, addition of strong authenticator or other reasons.
* Device's risk has changed due to installation of unapproved software, connection to insecure pheripheral device, encryption of data or other reasons.
* Any other subject's risk changes due to variety of reasons.

### Event Specific Claims {#risk-level-change-event-specific-claims}

> The following optional claims MAY be present in a Session Presented event:

Reviewer 2: Typo: "in a Session Presented event" (it should be "in a Risk Level Change event")

> 

Reviewer: Could we add a field that describes the type of risk? You mention in your example user risk and device risk. What if we had a field called `risk_type` whose values were an enum:
- DEVICE
- USER
- ACCOUNT
- OTHER

There may be other values that we should include in this enum as well.

Author: @FragLegs I thought we could drive the type using the subject claim. With a single enum it would be hard to flag a single risk event if it applies to more than 1 entity like USER + DEVICE etc.

Author: @tulshi I described that the existing subject of any CAEP event describes who this event is for, and it is clarified in [this](https://openid.github.io/sharedsignals/openid-sharedsignals-framework-1_0.html#name-subject-members) section. In other words, if the risk level change event has 2 subjects, a user and a device, then it should be perceived as risk being associated with both entities.

Reviewer 3: I guess my question is let's say there is a user X, who has a phone number Y. The transmitter wants to send out a risk level change about the user (and not their phone), but wants to describe the user using their phone number (i.e. the subject format is the ["phone_number"](https://www.rfc-editor.org/rfc/rfc9493.html#name-phone-number-identifier-for)), then how can the Transmitter convey this information?

Author: Thanks, @tulshi, for your comment.

I think this could also be a similar situation for session-related events.

e.g. There are separate sessions created with the user as a principal and the device as a principal.

> The transmitter wants to send out a session revoked/presented/established about the user (and not their device), but wants to describe the user using their phone number (i.e. the subject format is the ["phone_number"](https://www.rfc-editor.org/rfc/rfc9493.html#name-phone-number-identifier-for))

Reviewer: A `principal` claim works for me!

> reason_admin

Reviewer: I think it is a mistake to make this optional field required and to use it for passing important info about the event.

Author: @FragLegs @timcappalli we could define a new attribute that describes this, please let me know if you have any other suggestions

Author: Added `risk_reason` as a new field to describe the nature of the risk

: REQUIRED, JSON string indicates the reason that contributed to the risk level changes by the Transmitter. Overall this is an OPTIONAL claim, but marked as REQUIRED for Risk Level Change to indicate reasons behind the risk level change.

current_level 

> : REQUIRED, JSON string indicates the current level of the risk for the subject. Value MUST be one of LOW, MEDIUM, HIGH

Reviewer 2: Mandating the use of a LOW, MEDIUM, HIGH for risk levels may be too restricted for some usage. Some Companies may want to use a numeric range (e.g., 1-5 or 1-10) instead without needing to map it into LOW, MEDIUM, HIGH range.

Author: [Assurance level change](https://openid.net/specs/openid-caep-1_0-ID2.html#section-3.4.1-1.2.1) uses `namespace`, but in that example, there are standardized ways of representing levels.

In this case, I feel even the numeric system will be subjective, it will not help bring standardization. 

Another option is to plainly define a `score` as an optional attribute that could satisfy your use case, but I believe that could be solved by adopting https://github.com/openid/sharedsignals/issues/221

Reviewer 3: The [assurance level change](https://openid.net/specs/openid-caep-1_0-ID2.html#name-assurance-level-change) event has a "Any other value..." escape, that makes it possible to define your own namespace.  I feel having a standard set of values - whether Low, Medium, High or 1-10, will improve interoperability, so I'm in favor of keeping it simple. Like Apoorva says, we can include additional information in the "event data" field. I feel the numeric range is slightly more expressive, and you could define LOW, MEDIUM and HIGH as three points in that range, so it subsumes that.

* User's risk has changed due to potential suspecious access from unknown destination, password compromise, addition of strong authenticator or other reasons.
* Device's risk has changed due to installation of unapproved software, connection to insecure pheripheral device, encryption of data or other reasons.
* Any other subject's risk changes due to variety of reasons.

### Event Specific Claims {#risk-level-change-event-specific-claims}

risk_reason
: RECOMMENDED, JSON string: indicates the reason that contributed to the risk level changes by the Transmitter.

principal
: REQUIRED, JSON string: representing the principal entity involved in the observed risk event, as identified by the transmitter. The subject principal can be one of the following entities USER, DEVICE, SESSION, TENANT, ORG_UNIT, GROUP, or any other entity as defined in {{Section 2 of SSF}}. This claim identifies the primary subject associated with the event, and helps to contextualize the risk relative to the entity involved.

> current_level 

Reviewer 3: @appsdesh any thoughts on changing this to a numeric value on a fixed scale (e.g. 10)? We discussed this during the [SSWG meeting on 11/5](https://hackmd.io/@oidf-wg-sse/wg-meeting-20241105), but I don't remember if we agreed to switch to a numeric value. three of those values could map to "low", "medium" and "high" as you require, but it will give a little more expressivity.

Author: @tulshi welcome back! we already aligned on it in one of the comments [here](https://github.com/openid/sharedsignals/pull/205#discussion_r1843240302).

current_level 
: REQUIRED, JSON string: indicates the current level of the risk for the subject. Value MUST be one of LOW, MEDIUM, HIGH

previous_level 
: OPTIONAL, JSON string: indicates the previously known level of the risk for the subject. Value MUST be one of LOW, MEDIUM, HIGH. If the Transmitter omits this value, the Receiver MUST assume that the previous risk level is unknown to the Transmitter.

### Examples {#risk-level-change-examples}
The following is a non-normative example of a Risk Level Change event:

~~~json
{
    "iss": "https://idp.example.com/123456789/",
    "jti": "24c63fb56e5a2d77a6b512616ca9fa24",
    "iat": 1615305159,
    "aud": "https://sp.example.com/caep",
    "txn": 8675309,
    "sub_id": {
      "format": "iss_sub",
      "iss": "https://idp.example.com/3456789/",
      "sub": "jane.doe@example.com"
    },
    "events":{
      "https://schemas.openid.net/secevent/caep/event-type/risk-level-change":{

>          "current_level": "LOW",

Reviewer: Can we add "principal" to the example?

         "previous_level": "HIGH",
         "event_timestamp": 1615304991643,

>          "risk_reason":{

Reviewer: This should be a string, not be a dict of language to string
